### ISSUE TITLE
instruments/latents.py: capture the refinement trajectory on the production model

### ISSUE THREAD

**Issue by MatthewLacerda2** [MatthewLacerda2/TinyRefinementModel#225]
**The foundation for every depth instrument.** Nothing here is a research claim; it is the
plumbing that makes the claims measurable.

## The gap

`CausalRefiner` can already hand back every refinement state — `return_all_states` at
`trm/model/refiner.py:171`. The **production** model cannot:

```
trm/model/refiner.py         return_all_states: yes  (toy harness only)
trm/model/refiner_lm.py      return_all_states: no
trm/model/contract.py        return_all_states: no
trm/model/reasoner.py        return_all_states: no
```

So every question about what the refine loop actually does can be asked of a toy model at
dim96 and of nothing else. That is why depth inertness was found by reading eight sampled
prompts at the end of a 10-day run instead of measured at any point during it.

Two smaller problems in the same code:

- `return_all_states` is nested **inside** `return_all_iters` (`refiner.py:267`), so asking
  for states forces the full `[depth, b, s, vocab]` logits tensor — the thing the comment
  correctly calls unaffordable at real vocab. States alone are not expensive; the coupling is.
- The docstring says "toy scale only". For *logits* that is true. For *states* it is not, and
  the conflation is what has kept this off the production model.

## It is cheap

A full trajectory at production shape — depth 8, batch 1, seq 512, dim 960, f16:

```
9 states x 1 x 512 x 960 x 2 bytes = 8.8 MB
```

Eight megabytes. This has been affordable the entire time.

## What to build

1. Thread `return_all_states` through `contract.py` → `refiner_lm.py`, **decoupled** from
   `return_all_iters`, returning `[depth+1, b, s, dim]` including `z_0` (the encoder output,
   before any refinement — the trajectory's starting point, currently not returned at all).
2. `instruments/latents.py` — `capture(checkpoint, tokens, depth) -> Trajectory`, the one API
   every downstream instrument uses. Restores through `trm/runtime/restore.py` like the rest
   of `instruments/`.
3. The first-order readouts, which need nothing beyond the capture:
   - **step size** `‖z_k − z_{k−1}‖` per k — geometric decay means a fixed point; flat means
     the loop is still working
   - **turning angle** — cosine between successive steps; near +1 is walking a straight line,
     near −1 is oscillating, near 0 is wandering
   - **gate openness** per k (already computed inside the loop, currently discarded)
   - distance to the final state `‖z_k − z_K‖` — how many passes until it has arrived

## Why this first

Issues for the readout blind-spot probe and the trajectory visualiser both depend on this
capture existing. Depth-sensitivity logging does not, and can proceed in parallel.

Non-goal: this issue makes no claim about what the trajectories will show. It ships the
instrument, and the numbers above are descriptive, not a hypothesis under test.

## Lane

`trm/model/` (three small signature changes) + a new `instruments/latents.py`. Fully
CPU-testable — capture at toy dim in a unit test, assert shapes and that `z_0` is the encoder
output and `z_K` matches the ordinary forward's hidden state exactly.

**Comment by MatthewLacerda2** [comment]
Instrument built and pointed at the champion. It found something on the first run.

## The gate is data-dependent, and it explains the code result

Champion 3899391, depth 8, 16 held-out documents per corpus:

```
                  gate openness per pass (1 → 8)
fineweb-edu   0.186  0.00625 0.00625 0.00625 0.00625 0.00625 0.00625 0.00625
finemath      0.169  0.00625 0.00625 0.00625 0.00625 0.00625 0.00625 0.00625
codeparrot    0.233  0.113   0.103   0.100   0.104   0.108   0.108   0.105

                  step size ‖z_k − z_{k−1}‖
fineweb-edu   380  37  16   6   2.7  2.8  2.1  2.4
finemath      359  38  19   9   5.3  4.5  3.4  3.6
codeparrot    445 116  75  29  33   33   15   21

                  turning angle cos∠(step_k, step_{k−1})
fineweb-edu   -0.14 -0.85 -0.79 -0.65 -0.35 -0.02 -0.54
finemath      -0.08 -0.85 -0.80 -0.78 -0.63 -0.34 -0.46
codeparrot    -0.12  0.15  0.18 -0.46  0.34 -0.02 -0.18
```

Three things, and they agree:

1. **On prose and math the gate is saturated shut.** 0.00625 ≈ sigmoid(−5.07), and it is
   *bit-identical* across every pass and every document — so it is no longer input-dependent
   at all. On code it sits ~17× wider and keeps being modulated pass to pass.
2. **The steps collapse on prose and persist on code.** By pass 5 a prose trajectory has
   essentially arrived; a code trajectory is still moving an order of magnitude more.
3. **On prose, consecutive passes undo each other** (−0.85, −0.79, −0.65: near-reversals). On
   code they mostly advance.

## Why this matters

`gate_bias` initializes at **0.0**, i.e. sigmoid(0) = 0.5, a balanced gate. The trained model
sits at 0.006 on prose. **It learned to shut the loop off** — this is not an initialization
artifact.

And the split is exactly the one the CE measurements found: depth buys nothing on fineweb-edu
or finemath, and buys a real if small gain on codeparrot, concentrated on delimiters
(brackets + punctuation = 19% of tokens, 76% of the gain, scaling with nesting depth). We had
that result for two weeks with no mechanism. **The gate is the mechanism.** The model learned
where iterative refinement pays and turns it on there.

It also reframes #227's dichotomy — "the block is idle" versus "the readout cannot see it" —
because the answer is neither. The block is *gated*, deliberately and data-dependently. On
prose its output really is being discarded, but by a learned decision rather than a failure.

## What this is not yet

- **n = 16 documents per corpus, one checkpoint, no error bars.** The 17× is far larger than
  any plausible noise, and the prose value is a bit-identical constant rather than a noisy
  quantity, but this has not been replicated across the checkpoint ladder.
- **Gate openness here is a mean over positions and channels.** A mean of 0.006 could in
  principle hide a handful of positions with wide-open gates.

That second caveat is also the sharpest next measurement, and it is a real prediction: if the
gate is what selects for depth, and the code gain lives on delimiters, then **a per-position
gate map should open on brackets and stay shut on identifiers** — matching the 19%/76% split
already measured. If it does, the mechanism is nailed end to end. If the gate is uniformly
open across code tokens, then the gate explains code-vs-prose but *not* the within-code
attribution, and something else carries that.

The loop currently means the gate over positions before returning it, so this needs one more
small change to keep the per-position map.

## Notes for review

- Capture is a separate `capture_trajectory` method, **not** a flag threaded through
  `__call__` as this issue proposed. `contract.py` is what the training loop requires, and the
  loop never wants a trajectory; a flag there would sit on the jitted call the trainer runs
  millions of times. The default raises rather than returning a flat trajectory, so an
  instrument pointed at the non-recurrent control cannot quietly conclude that depth does
  nothing.
- The final captured state is **bit-identical** to the ordinary forward pass. Getting there
  needed a real fix: norming the stacked `[depth+1, b, s, dim]` array is exact on CUDA and
  **not** on the CPU emitter, because float ops are not shape-invariant (the #206 lesson).
  Each state is normed at its natural shape now, which is exact on both.

**Comment by MatthewLacerda2** [comment]
## Correction to the comment above, and a much sharper result

My previous comment described gate openness as a soft blend coefficient that sits ~17× wider
on code than prose. **That reading was wrong.** A peer session pushed back on the
bit-identical `0.00625` — asking whether it was a learned saturation or a clamp — and checking
it properly changed the finding.

It is neither. **The gate is binary.**

Pre-activations run to ±15,000–35,000, so every sigmoid is exactly 0.0 or exactly 1.0. That is
why the f32 mean and the "fraction above 0.5" agree to every printed digit:

```
                 g mean (f32)    frac(g > 0.5)
fineweb-edu        0.006250        0.00625
codeparrot         0.085536        0.08554
```

`0.00625 = 1/160`, and at `dim = 960` that is **exactly 6 channels**. The number was not a soft
gate value at all — it was a *count*.

## What the loop actually does

```
fineweb-edu, passes 2-8   strictly 0/1: True    open channels per position: min 6, max 6
                          open somewhere: 6     open everywhere: 6 of 960
                          channel ids: [47, 417, 554, 603, 788, 910]

codeparrot,  passes 2-8   strictly 0/1: False   open channels per position: min 6, max 447
                          open somewhere: ~480  open everywhere: 4 of 960
```

**On prose the refine loop writes to 6 of 960 channels — the same six, at every position, on
every pass.** It is not "doing less work"; it is structurally incapable of content-dependent
work, because the routing does not depend on the content. 0.625% of the state is live and the
other 99.375% is a hard identity passthrough.

On code the same loop opens between 6 and 447 channels **depending on the position**.

## Which positions

Pass 3, 12 documents, mean open channels by token category:

```
codeparrot                                  fineweb-edu
  whitespace        236.0   (37.6%)           punctuation        6.0
  word/identifier     6.0   (38.4%)           word/identifier    6.0
  punctuation         6.0   (10.7%)           whitespace         6.0
  bracket             6.0    (3.0%)           bracket            6.0
  number              6.0    (1.6%)           number             6.0
```

**The gate opens on whitespace in code, and on nothing at all in prose — whitespace
included.** In Python, whitespace is newlines and indentation: it *is* the block structure. In
prose it separates words and carries nothing. The gate opens exactly on the tokens that carry
structural information, and only in the corpus where they do.

This is a mechanism for the two-week-old result that depth helps on code and nowhere else, and
it is consistent with the earlier conclusion that depth is doing state tracking over nested
structure — indentation *is* the nesting in Python.

## What I am NOT claiming

I predicted the gate would open on **brackets**, since brackets plus punctuation are 19% of
code tokens and carry 76% of the measured CE gain. **It does not.** Brackets get the same 6
channels as everything else. So the gate story and the delimiter-attribution story do not yet
join up, and there is a specific reason they might only look disjoint:

**an alignment question I have not resolved.** Refinement at position *i* improves the
prediction of token *i+1*. The delimiter attribution bucketed by the category of the *target*
token; this buckets by the category of the position where the gate opened. "Gate opens on
whitespace" and "gain appears on brackets" are compatible if brackets tend to *follow*
whitespace — which in indented code they very often do. That is one shift of an index away
from being settled, and it should be settled before either result is written up.

Other limits: 12–16 documents, one checkpoint, no error bars, and no replication across the
0.78B/1.40B/3.20B ladder.

## A numerical problem worth its own issue

Gate pre-activations reach ±35,000 and **overflow to `inf` on codeparrot pass 1** under the f16
compute policy (f16 max is 65,504). Two consequences:

1. A saturated sigmoid has exactly zero gradient. **The gate can no longer learn** — it froze
   into whatever routing it had when the pre-activations blew up, and no amount of further
   training moves it.
2. It is a live candidate for #229, where this architecture returns all-NaN logits on ~10% of
   code documents at 3.20B tokens. An overflowing gate pre-activation is on the code path, on
   the corpus where it happens, and produces whole-window contamination — which matches #229's
   measured signature exactly (13/300 documents non-finite at *every* position).

Filing separately.
